libs/aggregation_example/src/avg.rs

```diff
@@ -1,29 +1,159 @@
+    let logfile = FileAppender::builder()
+        // Pattern: https://docs.rs/log4rs/*/log4rs/encode/pattern/index.html
+        .encoder(Box::new(PatternEncoder::new("{l}: {m}\n")))
+        .append(false)
+        .build("sim.log")
+        .unwrap();
+    // Log Trace level output to file where trace is the default level
+    // and the programmatically specified level to stderr.
+    let config = Config::builder()
+        .appender(Appender::builder().build("logfile", Box::new(logfile)))
+        .appender(
+            Appender::builder()
+                .filter(Box::new(ThresholdFilter::new(log::LevelFilter::Info)))
+                .build("stderr", Box::new(stderr)),
+        )
+        .build(
+            Root::builder()
+                .appender("logfile")
+                .appender("stderr")
+                .build(log::LevelFilter::Trace),
+        )
+        .unwrap();
+    // Use this to change log levels at runtime.
+    // This means you can change the default log level to trace
+    // if you are trying to debug an issue and need more logs on then turn it off
+    // once you are done.
+    let _handle = log4rs::init_config(config);
+}
+
+// user defined functions:
+
+// udf_type: Aggregation
+// init_func: new
+// exec_func: execute
+// return_func: get_avg
+// struct_name: Avg
+// id: avg
+
+#[derive(Clone, Copy, Debug)]
+pub struct Avg {
+    avg: u64,
+    total: u64,
+    num_instances: u64,
+}
+
+impl Avg {
+    fn new() -> Avg {
+        Avg {
+            avg: 0,
+            total: 0,
+            num_instances: 0,
+        }
+    }
+    fn execute(&mut self, _trace_id: u64, instance: String) {
+        self.total += instance.parse::<u64>().unwrap();
+        self.num_instances += 1;
+    }
+    fn get_avg(&mut self) -> String {
+        self.avg = self.total / self.num_instances;
+        self.avg.to_string()
+    }
+}
+
 #[derive(Clone, Debug)]
-pub struct AggregationStruct {
-    pub count: u32,
-    pub total: u32,
+pub struct Filter {
+    avg: Avg,
 }
 
-impl AggregationStruct {
+impl Filter {
     #[no_mangle]
-    pub fn input(&mut self, input_string: String) {
-        let val = input_string.parse::<u32>().unwrap();
-        self.total += val;
-        self.count += 1;
+    pub fn new() -> *mut Filter {
+        log_setup();
+        Box::into_raw(Box::new(Filter { avg: Avg::new() }))
     }
 
     #[no_mangle]
-    pub fn return_value(&mut self) -> String {
-        if self.count == 0 {
-            return "0".to_string();
-        }
-        let avg = self.total / self.count;
-        return avg.to_string();
+    pub fn new_with_envoy_properties(string_data: IndexMap<String, String>) -> *mut Filter {
+        log_setup();
+        Box::into_raw(Box::new(Filter { avg: Avg::new() }))
+    }
+
+    pub fn on_incoming_requests(&mut self, mut x: Rpc) -> Vec<Rpc> {
+        self.avg.execute(x.uid, x.data.clone());
+        let mut rpc_str = "avg: ".to_string();
+        rpc_str.push_str(&self.avg.get_avg());
```

Comment:
I would just merge get_avg into execute. It is the same process as far as I understand.